File: 2020/day03.jl
input=permutedims(stack(readlines("data/input03.txt")))

function part1(input,(slope1,slope2))
    x,y=1,1
    s=0
    while y<size(input,1)
        x+=slope1
        y+=slope2
        x=mod1(x,size(input,2))
        if input[y,x] == '#'
            s+=1
        end
    end
    return s
end

function part2(input)
    slopes=[(1,1),(3,1),(5,1),(7,1),(1,2)]
    p=1
    for slope in slopes
        p*=part1(input,slope)
    end
    return p
end

@time println(part1(input,(3,1)))
@time println(part2(input))

File: 2020/day03.rs
use std::fs;
use std::io::{self, prelude::*};
use std::time::Instant;

fn read_input(file_path: &str) -> Vec<Vec<char>> {
    let file = fs::File::open(file_path).expect("Unable to open the file");
    let reader = io::BufReader::new(file);
    let mut lines = vec![];
    for line in reader.lines() {
        lines.push(line.unwrap().chars().collect());
    }
    lines
}

fn part1(input: &Vec<Vec<char>>, slope1: usize, slope2: usize) -> u32 {
    let mut x = 0;
    let mut y = 0;
    let mut trees_count = 0u32;
    while y < input.len() {
        if input[y][x] == '#' {
            trees_count += 1;
        }
        x = (x + slope1) % input[0].len();
        y += slope2;
    }
    trees_count
}

fn part2(input: &Vec<Vec<char>>) -> u32 {
    let slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    let mut result = 1u32;
    for &(slope1, slope2) in &slopes {
        result *= part1(input, slope1, slope2);
    }
    result
}

fn main() {
    let input = read_input("data/input03.txt");
    
    let start_time = Instant::now();
    let part1_result = part1(&input, 3, 1);
    let duration = start_time.elapsed();
    println!("Part 1: {} (Time elapsed: {:?})", part1_result, duration);
    
    let start_time = Instant::now();
    let part2_result = part2(&input);
    let duration = start_time.elapsed();
    println!("Part 2: {} (Time elapsed: {:?})", part2_result, duration);
}
